### ISSUE TITLE
[FEA] Write shuffle data in a background thread with flow control

### ISSUE THREAD

**Issue by revans2** [NVIDIA/spark-rapids#11341]
**Is your feature request related to a problem? Please describe.**
Ideally we want to be able to do I/O in separate threads from computation so that we can overlap computation with I/O and reduce the likelihood that we have to release the semaphore.

Currently when we do a shuffle write the data is pulled by the GpuShuffleExec and once the batch of GPU data is brought to the GPU the semaphore is released. This can cause a lot of problems with memory pressure on the GPU because the task writing the data might have some spillible memory on the GPU, and when we release the semaphore for the write other tasks get on the GPU and can cause that first task to spill.

Ideally we want to never release the semaphore until the task is complete (assuming that I/O is fast enough to keep up with that).

We want to have a separate writer thread for each task thread. This thread would be part of the SQL plugin and not the shuffle plugin so that we can have this advantage in all environments, and eventually for other types of writes. There would be a queue in between these two threads so that data produced by the task can be buffered. There would also be a coordinator that would monitor all of these queues to make sure that we don't go over a maximum number of bytes in flight.  

This bytes in flight number would be in addition to whatever number of bytes in flight shuffle itself might have. The config would also have to be a soft limit, as we need to be able to put at least one entire batch into the queue. I don't think we want to go over that limit in other cases, as it really is just about flow control.  That bytes in flight number should be for the entire process and not for an individual task.

So to get an idea of how this would work lets take a look at a simple example.

1. The writer thread will ask if there is anything more to process.  The writer thread will then block waiting on something from the queue.
2. The computation thread will then come up and start to process the task normally. As soon as it is ready with one batch of output it will ask the writer coordinator if there is room to put this batch into the queue.
3. The coordinator says there is room so the computation thread will copy the data to the CPU and put it into its queue. (note that we might be able to do an async copy/free and continue with the computation instead of waiting for it to complete). Note that the GPU semaphore is not released here.
4a. The writer thread wakes up and starts to process the batch (hands it to the shuffle). Once the shuffle iterator comes back and asks for another batch we tell the coordinator that the first batch is done so we can decrement the bytes in flight. (not sure if we want to wait for the shuffle to complete or if we decrement it right before we hand it to shuffle.
4b. At the same time GPU starts to work on the next batch. Lets say that this is from a window with a sort so the batch is already in GPU memory and it is ready very quickly.
5. The computation thread then does the same thing it asks if there is room to put it into the queue.
6. The coordinator says yes and it is put into the queue, but the shuffle thread is not done yet.
7. The computation thread the copies the data to the CPU and puts it into the queue. (no semaphore is released)
8. The computation thread now computes the next batch of data and asks the coordinator if there is room for it.
9. The coordinator says that there is no room for it, so the GPU blocks waiting for the coordinator to tell it that there is more room. (note that the Semaphore has not been released)
10. The shuffle finally finishes with the first batch and asks for another batch.
11. The writer thread pulls the next batch from the queue and tells the coordinator that there is more room.
12. The coordinator then looks at the pending requests along with the priority of each task and picks a task to wake up. (for now I think it should be a strict priority order, but we could play games with letting small things in).

The process then continues until all of the input batches are complete.

Note also that because we are not releasing the semaphore as often as before there is a higher probability that we will do I/O with the semaphore held. We need to do a lot of profiling to see if we want to also update the shuffle, parquet, orc, etc reader code to release the semaphore before doing I/O. This might also need to be under a config in the short term and we might see a regression in some cases.

This is where we want to be, so the config would be short term if we do see any regressions.

**Comment by abellina** [comment]
@sperlingxx could you take a look at this? I would also add that this task should start with some sort of benchmark (contributing to or closing https://github.com/NVIDIA/spark-rapids/issues/11575).

**Comment by revans2** [comment]
@sperlingxx 

At some point the coordinator is likely going to take on more work and merge with state machine and budgeting feedback loops in the SparkResourceAdaptor.  For now it is just going to deal with shuffle data and async file write data. When we support async file reads, then we will update it for that too. When we try to do CPU UDF like computation in a separate thread pool, then we can also add in host memory budgeting there too when the time comes.

[spark.rapids.sql.asyncWrite.maxInFlightHostMemoryBytes](https://github.com/NVIDIA/spark-rapids/blob/e5c64f6b4d10812745d5f12eefb8175d412594a6/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala#L2536-L2543) was added as a part of https://github.com/NVIDIA/spark-rapids/pull/11730 to let us limit the amount of host memory used when writing output files in a background thread. Currently that is done by [`com.nvidia.spark.rapids.io.async.TrafficController`](https://github.com/NVIDIA/spark-rapids/blob/branch-25.06/sql-plugin/src/main/scala/com/nvidia/spark/rapids/io/async/TrafficController.scala). 

The controller I spoke about here should do effectively the same job. We have a host memory budget that we can use to try and do other things (like I/O) in background threads so that a single large task could be doing multiple things at once. We will worry about CPU and GPU computation resources separately. 

How do we divide that memory budget up between tasks, or between different parts of a task? That is something we do not totally know yet and we need to do experimentation to find what is good enough in the short term. Right now the idea is to have a strict priority order https://github.com/NVIDIA/spark-rapids/issues/12472 between tasks. The important thing is that we cannot deadlock/livelock. 

There are several different ways that we can deal with this. 

We can have the coordinator know about different parts of a task's processing and determine when it needs to go over the limit (which could cause spilling to happen) so that we don't deadlock. Effectively it would be when all of the memory budget is used by early stage operations (like shuffle reads) and all later stage operations (like shuffle writes) are blocked.

We could partition the budget between different stages within a task. I.e. writes get 1/2 the budget and reads get 1/2 the budget. This works okay, but we probably would want to have follow on work to dynamically adjust the budgets based on monitoring the usage of the budgets to optimize the flow of the data. In either case we would have to guarantee that each stage could hold at least one batch of data. This does get a little more complicated if we try to add in column to row and row to column processing, but it still should be doable. If we do divide up the budget we have to pay attention to what happens if there are multiple inputs, like a join.

There are other ways too, and I am open to other ideas. But these are the ones I could think of right now that do not have too many downsides.

We may also need to coordinate a bit with the SparkResourceAdaptor in case a thread becomes blocked. This is especially true for the case when the coordinator is monitoring the task's data to see if it is deadlocked or not. Having a single task priority should make this mush less likely to happen, but it does not make it impossible.

> When does the computation thread submit the async shuffle write task? And be able to start to process the next batch. I think the handover point should be after the D2H memory copy, so as not to keep two batches in device. But if so, is it probable to lead to enormous host memory pressure?

This is where flow control is going to help us. I think we want each sub section of a task will try to reserve room for its output before it does anything.

Lets walk through a simple example here.

shuffle read -> GPU computation -> shuffle write

We want three separate threads. One for shuffle read, one for shuffle write, and one for the GPU computation.

The shuffle read thread will start off by asking the coordinator if there is room for 1 target batch size of data. If there is no room, the thread will block until the coordinator says that there is now room. When there is room, the thread will start to read the data. Once the data is buffered enough that it has about 1 target batch size worth of data it will insert it into the queue, update the coordinator with the amount that was actually read and ask if there is room for another one. This is just going to repeat until all of the data is read or the task exits, which can happen on operations like limit.

The GPU computation thread is a little more complicated because it uses iterators to deal with the complex processing that could happen. It will start off by asking the coordinator if there is room for 1 target batch size of data for its output. If there is no room, it will block until the coordinator says that there is room. When there is room, it will pull a batch of output from the iterator, put it into its output queue, update the coordinator with the size that was actually inserted, and then repeat until all of the data has been processed or the task completes. The iterators that provide input to the GPU computation thread will pop batches off of the queues coming from shuffle. When a batch is popped off, the coordinator is updated that it was removed so that it can then wake up other threads to do more work. If there is nothing in the queue and we don't know if there is more data that could show up, then we need to inform the coordinator before it blocks. This is so that the coordinator can decide if it should wake up an upstream thread to do more processing and possibly go over the budget, or if it is okay for this to just block. 

The shuffle writer thread is then not checking about output budgets directly so it does not need to ask the coordinator for a budget. It is just popping off the partitioned batches from its queue and handing them to the shuffle service as fast as it can go.

In the future we may add in the ability to know if we are using multi-threaded shuffle or not and combine things so we don't need as many threads. For now I just want this to be agnostic of the shuffle implementation being used.

Should we place partitioned CPU data to the queue for the shuffle write service or partitioned GPU data to it? I think we want to hand CPU data to it, because it is likely to be a more common resource. I can see situations where we have a very small CPU setup and a very large GPU. If we do start to see a lot of these we can adjust, but for now I think we want to copy the data to the CPU before we put it into the queue.

**Comment by sperlingxx** [comment]
Since the async shuffle write should "be agnostic of the shuffle implementation being used", it seems that we cannot simply  replace current pull-based iterator model with some kind of push-based execution model because spark-rapids cannot determine the behavior of `ShuffleWriter`.

Instead, we might regard "The GPU computation thread" as an async provider, which is connected to the main thread of Spark Task by a "queue".  The GPU computation thread produces partitioned HostVectors with corresponding slice functor: 
* If slice happens on CPU, the slice functor is pretty much a wrapper of `sliceInternalOnCpuAndClose`.
* If slice happens on GPU, the slice functor is just a lightweight wrapper which bypasses the actual data processing. Since the actual slice has been applied by "The GPU computation thread".

We will have something like `GpuComputeRdd` which delegates the RDD involving every GPU operations. When the `next` method being called, the iterator of `GpuComputeRdd` tries to poll element from the "queue". And  the "queue" is filled by an independent producer thread which exhausts the delegated RDD continuously unless being blocked by "the `coordinator`".

**Comment by revans2** [comment]
@sperlingxx 

1. I am fine with a size-limited blocking queue in the first version. We have to have it be a soft limit where if there is nothing in the queue that we can still add at least one batch. Longer term we have to deal with the complex issues.
2.  I think that the answer is going to be both yes and no.

Let me explain. 

There are a few different situations that we can be in.

1. We are GPU computation bound
2. We are CPU computation bound
3. We are I/O bound on disk
4. We are I/O bound on network
5. We are constrained by working memory on the GPU
6. We are constrained by working memory on the CPU.

We can dynamically switch between which one is blocking our computation at any point in time.

So for a shuffle write. Lets say that we are bound by disk I/O writing shuffle data out.  If we add in another GPU thread it is not going to speed anything up. It is just going to increase memory pressure on the GPU and possible CPU, which might cause us to spill and end up slowing down the even more, because the disk is the bottleneck.

Now lets say that we are bound by single threaded CPU compression instead of disk I/O. In this case releasing the semaphore and allowing another task onto the GPU might help, because it could allow another CPU thread to run compression without interfering with  the first, so long as we have enough memory to handle it without spilling, If we don't we might slow things down because we end up spilling too much from the first task.

Now lest say that we are bound by CPU compression instead of disk I/O, but we are using the multi-threaded shuffle and we are using all of the CPU cores that are available. Now adding another task to the GPU is not likely to speed anything up.

My long term goal is to effectively have a queue in between each stage of the processing within a single task and that anything that can be parallelized will be.  That way if we have a large task, think 20 batches of data, then some threads will be reading input data, which is I/O intensive, while other threads are processing some of it on the GPU, while at the same time other threads are compressing/encrypting the data in a thread pool, and then finally other threads are writing it out to disk.

When all of that is done, then we can monitor the utilization of each of the resources and make intelligent decisions on how to divide up the resources.

Like If we see that we are I/O bound on disk IOps, not bandwidth, then we can  try to give the shuffle algorithm more memory so it can hopefully coalesce more shuffle data into larger pieces and improve the disk bandwidth.

If we see that we are GPU bound, and we have free GPU memory we might try and increase the batch size to process the data more efficiently. If there is no more memory or the batch size is already large, we probably would just do nothing because there is no point to try and add more to the GPU, it is full. There is little point in doing more I/O because it will not speed up the GPU.

Without that extra knowledge we are just guessing as to what to do.

**Comment by sperlingxx** [comment]
I am trying to do some preparation work at first: #12819
